### PR TITLE
feat(core): separate refs from vars.

### PR DESCRIPTION
**Pull request by tbosch**
Does not need to go into G3...

### CODE REVIEW

File: modules/angular2/src/compiler/template_parser.ts [hunk 1]
```diff
@@ -431,46 +451,68 @@ class TemplateParseVisitor implements HtmlAstVisitor {
...
-        this._parseProperty(bindParts[5], attrValue, attr.sourceSpan, targetMatchableAttrs,
+        this._parseProperty(bindParts[7], attrValue, attr.sourceSpan, targetMatchableAttrs,
                             targetProps);
 
-      } else if (isPresent(
-                     bindParts[2])) {  // match: var-name / var-name="iden" / #name / #name="iden"
-        var identifier = bindParts[5];
-        this._parseVariable(identifier, attrValue, attr.sourceSpan, targetVars);
+      } else if (isPresent(bindParts[2])) {  // match: var-name / var-name="iden"
+        this._reportError(`"var-" is deprecated. Use "ref-" or "let-" instead!`, attr.sourceSpan,
```
Comment (tbosch) [inline modules/angular2/src/compiler/template_parser.ts]: Report the exact replacement, as we know it by checking the `isTemplateElement` flag

File: modules/angular2/src/compiler/view_compiler/view_builder.ts [hunk 2]
```diff
@@ -72,7 +73,7 @@ export function buildView(view: CompileView, template: TemplateAst[],
                                                  view.declarationElement :
                                                  view.declarationElement.parent);
   // Need to separate binding from creation to be able to refer to
-  // variables that have been declared after usage.
+  // locals that have been declared after usage.
```
Comment (tbosch) [inline modules/angular2/src/compiler/view_compiler/view_builder.ts]: references

File: modules/angular2/src/compiler/runtime_compiler.ts [hunk 3]
```diff
@@ -30,7 +30,7 @@ import {
   NgContentAst,
   EmbeddedTemplateAst,
   ElementAst,
-  VariableAst,
+  ReferenceAst,
```
Comment (tbosch) [inline modules/angular2/src/compiler/runtime_compiler.ts]: Not used, remove

File: modules/angular2/src/compiler/template_parser.ts [hunk 4]
```diff
@@ -122,15 +129,20 @@ export class TemplateParseResult {
...
   parse(component: CompileDirectiveMetadata, template: string,
         directives: CompileDirectiveMetadata[], pipes: CompilePipeMetadata[],
         templateUrl: string): TemplateAst[] {
     var result = this.tryParse(component, template, directives, pipes, templateUrl);
-    if (isPresent(result.errors)) {
-      var errorString = result.errors.join('\n');
+    var warnings = result.errors.filter(error => error.level === ParseErrorLevel.WARNING);
+    var errors = result.errors.filter(error => error.level === ParseErrorLevel.FATAL);
+    if (warnings.length > 0) {
+      this._console.log(`Template parse warnings:\n${warnings.join('\n')}`);
```
Comment (tbosch) [inline modules/angular2/src/compiler/template_parser.ts]: Use `DOM.logWarning`

File: modules/angular2/src/compiler/template_parser.ts [hunk 5]
```diff
@@ -488,11 +530,21 @@ class TemplateParseVisitor implements HtmlAstVisitor {
   }
 
   private _parseVariable(identifier: string, value: string, sourceSpan: ParseSourceSpan,
-                         targetVars: VariableAst[]) {
+                         targetVars: VariableAst[], targetMatchableAttrs: string[][]) {
```
Comment (tbosch) [inline modules/angular2/src/compiler/template_parser.ts]: Remove matchable attrs completely, (also from within expressions)

File: modules/angular2/src/compiler/template_parser.ts [hunk 6]
```diff
@@ -547,55 +599,58 @@ class TemplateParseVisitor implements HtmlAstVisitor {
...
+    });
+    return directives.filter(dir => isPresent(dir));
+  }
+
+  private _createDirectiveAsts(isTemplateElement: boolean, elementName: string,
+                               directives: CompileDirectiveMetadata[],
                                props: BoundElementOrDirectiveProperty[],
-                               possibleExportAsVars: VariableAst[],
-                               sourceSpan: ParseSourceSpan): DirectiveAst[] {
+                               possibleReferences: ElementOrDirectiveRef[],
```
Comment (tbosch) [inline modules/angular2/src/compiler/template_parser.ts]: should be `elementOrDirectiveRefs`

File: modules/angular2/src/compiler/template_parser.ts [hunk 7]
```diff
@@ -547,55 +599,58 @@ class TemplateParseVisitor implements HtmlAstVisitor {
...
-                              exportAsVars, sourceSpan);
+                              sourceSpan);
     });
-    possibleExportAsVars.forEach((varAst) => {
-      if (varAst.value.length > 0 && !SetWrapper.has(matchedVariables, varAst.name)) {
-        this._reportError(`There is no directive with "exportAs" set to "${varAst.value}"`,
-                          varAst.sourceSpan);
+    possibleReferences.forEach((elOrDirRef) => {
+      if (elOrDirRef.value.length > 0) {
+        if (!SetWrapper.has(matchedVariables, elOrDirRef.name)) {
```
Comment (tbosch) [inline modules/angular2/src/compiler/template_parser.ts]: `matchedReferences`

File: modules/angular2/test/compiler/template_parser_spec.ts [hunk 8]
```diff
@@ -906,6 +971,15 @@ There is no directive with "exportAs" set to "dirA" ("<div [ERROR ->]#a="dirA"><
                 ]);
           });
 
+          it('should locate directives in references', () => {
```
Comment (tbosch) [inline modules/angular2/test/compiler/template_parser_spec.ts]: Should not locate directives in references

Comment (tbosch) [inline modules/angular2/test/compiler/template_parser_spec.ts]: Same for `#` and inside of expressions

**Comment by tbosch** [comment]
Reviewd in person with @mhevery

File: modules/angular2/src/compiler/view_compiler/view_builder.ts (hunk 2 above)
Comment (tbosch) [inline modules/angular2/src/compiler/view_compiler/view_builder.ts]: done

File: modules/angular2/src/compiler/runtime_compiler.ts (hunk 3 above)
Comment (tbosch) [inline modules/angular2/src/compiler/runtime_compiler.ts]: Done

File: modules/angular2/src/compiler/template_parser.ts (hunk 4 above)
Comment (tbosch) [inline modules/angular2/src/compiler/template_parser.ts]: done

File: modules/angular2/src/compiler/template_parser.ts (hunk 5 above)
Comment (tbosch) [inline modules/angular2/src/compiler/template_parser.ts]: done

File: modules/angular2/src/compiler/template_parser.ts (hunk 6 above)
Comment (tbosch) [inline modules/angular2/src/compiler/template_parser.ts]: done

File: modules/angular2/src/compiler/template_parser.ts (hunk 7 above)
Comment (tbosch) [inline modules/angular2/src/compiler/template_parser.ts]: done

File: modules/angular2/test/compiler/template_parser_spec.ts (hunk 8 above)
Comment (tbosch) [inline modules/angular2/test/compiler/template_parser_spec.ts]: done

File: modules/angular2/src/compiler/template_ast.ts [hunk 9]
```diff
@@ -82,7 +82,18 @@ export class BoundEventAst implements TemplateAst {
...
+export class ReferenceAst implements TemplateAst {
+  constructor(public name: string, public value: CompileTokenMetadata,
+              public sourceSpan: ParseSourceSpan) {}
+  visit(visitor: TemplateAstVisitor, context: any): any {
+    return visitor.visitReference(this, context);
+  }
+}
+
+/**
+ * A variable declaration on a <template> (e.g. `let var="expression"`).
```
Comment (gkalpak) [inline modules/angular2/src/compiler/template_ast.ts]: _let var_  -->  _let-var_ ?

Comment (tbosch) [inline modules/angular2/src/compiler/template_ast.ts]: Thanks, yes.

File: modules/angular2/src/core/change_detection/change_detector_ref.ts [hunk 10]
```diff
@@ -70,7 +70,7 @@ export abstract class ChangeDetectorRef {
    * @Component({
    *   selector: 'giant-list',
    *   template: `
-   *     <li *ngFor="#d of dataProvider.data">Data {{d}}</lig>
+   *     <li *ngFor="let d of dataProvider.data">Data {{d}}</lig>
```
Comment (gkalpak) [inline modules/angular2/src/core/change_detection/change_detector_ref.ts]: _</lig>_  -->  _</li>_

**Pull request closed without merging**